### DISCUSSION TITLE
Lolin/wemos & i2C

### DISCUSSION THREAD

**Discussion by kjm1102** [micropython/micropython#11998] [Hardware & Peripherals]
I'm having trouble understanding the placement of pull up resistors on the D1 mini & the C3 pico. Both boards bring out the i2c bus on a dedicated 4 pin socket which would be super convenient if only it had pull resistors.

The  D1 mini, https://www.wemos.cc/en/latest/_static/files/sch_d1_mini_v4.0.0.pdf, has the i2c socket on io4 & i05 but the likely pullups are on io0 & io2. The C3 pico, https://www.wemos.cc/en/latest/_static/files/sch_c3_pico_v1.0.0.pdf, has the i2c socket on io8 & io10. There is a pullup for SDA on io8 but the likely SCL pullup is on io2 not i010.

Since the i2c socket is hardwired to the io pins SCL & SDA can't be re-assigned & I'm left scratching my head as to why the i2c pullups seem to be on the wrong pins?

**Comment by mcauser** [comment]
On the ESP8266 (D1 Mini v4.0.0), IO0 and IO2 are bootstrapping pins. They have to be pulled high to boot. Those 12k resistors aren't I2C pull-ups.
See https://randomnerdtutorials.com/esp8266-pinout-reference-gpios/

Judging by the schematics, IO4/IO5 are missing onboard I2C pull-ups. I assume this was on purpose, as most I2C peripherals have their own pull-ups on board and by not having them on the D1 mini means the pins can be used as plain old GPIO for any purpose.

The pin order on the Lolin I2C connector is also different from Stemma/QT.

Pin | Lolin | Stemma/QT
-- | -- | --
1 | GND | GND
2 | SDA | 3V3
3 | SCL | SDA
4 | 3V3 | SCL

Same applies for the other boards with onboard I2C socket, D1 mini pro, D32 pro, S2 pico etc. They all have the same I2C socket pin order.

S2 pico does have onboard I2C pull-ups, as there is an OLED onboard on the bus. The same bus exposed by the Lolin I2C connector.

On the C3 pico, you're right. The schematics show 10k pull-ups on IO8/IO2, but the I2C connector uses IO8/IO10.
Looking at the C3 datasheet, IO2, IO8 and IO9 are strapping pins.
https://www.espressif.com/sites/default/files/documentation/esp32-c3_datasheet_en.pdf
The 10k pull-up on IO8/SDA is not an I2C pull-up, it's to disable printing rom messages on boot.
This is why we don't also see a 10k on IO10.

**Reply by kjm1102** [reply]
Tnx Mike. Do you know why the 10k SDA/SCL pullups tend to be on the peripherals rather than the esp32 board? I thought the idea of I2C was that you can have multiple peripherals on the same bus? If each peripheral has it own pull ups then with 2 peripherals won't the net pullup value be 5k instead of 10k?

**Comment by robert-hh** [comment]
> Do you know why the 10k SDA/SCL pullups tend to be on the peripherals rather than the esp32 board? 

I guess that this is for the convenience of people using breakout boards. If one designs the an own board with various I2C devices, there would most probably only one pair of pull.up resistors on the board. With 10k pull-up at boards, you can connects quite a few of them in parallel before the impedance get's too low. The ESP32 can sink up to 40mA, so it could drive ~100 boards with 10k pull-ups. 

Not having pull-up resistors at the MCU board is good as well, because these would impede using the pins for other purposes. Unless the pull-up can be switched off, like at the PYBD boards.

**Reply by kjm1102** [reply]
If I could find a decent DC to DC with integrated inductor for 3v3 out from a lipo I'd be tempted to have a crack at board layout. I'd use passives that I could at least see with the naked eye!

**Reply by robert-hh** [reply]
There are step-up/step-down regulator breakout  boards from Pololu at about the size of a TO220 package. See the S9V11 or S7V8 series. Very handy. e.g. https://www.pololu.com/product/2122

**Reply by kjm1102** [reply]
nice
